### PR TITLE
C#: New query cs/possibly-incorrect-variable

### PR DESCRIPTION
**Pull request by calumgrant**
Finds some copy-and-paste bugs of the form
```
m_X = x;
m_Y = x;
```
or
```
X = m_x;
X = m_y;
```
and other variations.  See https://jira.semmle.com/browse/ODASA-5485?filter=13708

### CODE REVIEW

File: csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.qhelp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+</recommendation>
+
+<example>
+<p>The following example shows a class <code>Point1</code> where the constructor does not initialize the
+field <code>Y</code>:</p>
+
+<sample src="PossiblyIncorrectVariableBad.cs" />
+
+<p>
+The fix is to edit the assignment to initialize the field <code>Y</code> instead:
```
Comment (jf205) [inline csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.qhelp]:
```suggestion
To fix this, edit the assignment to initialize the field <code>Y</code> instead:
```

File: csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.qhelp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+</p>
+
+<sample src="PossiblyIncorrectVariableGood.cs" />
+
+</example>
+<references>
+
+  <li>Mitre: <a href="https://cwe.mitre.org/data/definitions/697.html">CWE-697: Incorrect Comparison</a></li>
```
Comment (jf205) [inline csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.qhelp]: This reference is exactly the same as the one automatically added in the rendered query help: https://jenkins.internal.semmle.com/job/Pull-Request-Check/job/QHelp-Preview/2627/artifact/out/ql/csharp/ql/src/Likely%20Bugs/PossiblyIncorrectVariable.html

Could you link to something else? (Or just remove it).

File: csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  Access getLeft() { result = this.getChild(0) }
+
+  Access getRight() { result = this.getChild(1) }
+}
+
+/**
+ * Holds if two comparisons or assignments are adjacent, either within an expression
+ * or in a sequence of assignments or declarations.
+ */
+predicate adjacent(ComparisonOrAssignment a1, ComparisonOrAssignment a2) {
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql]: With my suggestion above, this predicate is replaced by `a2 = a1.getNext()`.

File: csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  }
+}
+
+/*
+ * Holds for sequences of expressions of the form
+ * ```
+ * same1 = different1;
+ * same2 = different2;
+ * ```
+ * where the target of `same1` and `same1` is equal.
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql]: the target of `same1` and `same2`

File: csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  same1.getTarget() = same2.getTarget()
+}
+
+predicate duplicatedAccess2(Access same1, Access same2, Access different1, Access different2) {
+  duplicatedAccess1(same1, same2, different1, different2) and
+  any(SameVariables same).same(same1, same2) and
+  exists(Declaration target | target = same1.getTarget() |
+    not target.(Modifiable).isStatic() and
+    not target instanceof ImplicitAccessorParameter and
+    not target instanceof EnumConstant
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql]: How about moving these restrictions into `ScopedDeclaration` and replace `exists(Declaration target` with `exists(ScopedDeclaration target`?

File: csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    different1 = op1.getRight() and
+    same2 = op2.getLeft() and
+    different2 = op2.getRight()
+    or
+    same1 = op1.getRight() and
+    different1 = op1.getLeft() and
+    same2 = op2.getRight() and
+    different2 = op2.getLeft()
+  ) and
+  same1.getTarget() = same2.getTarget()
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql]: Could add `.(ScopedDeclaration)` to make this predicate smaller.

File: csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    candidateDecl.getScope() = same1.getTarget().(ScopedDeclaration).getScope() and
+    candidateDecl.getName().toLowerCase() = candidate
+  )
+}
+
+from Access same1, Access same2, Access different1, Access different2, ScopedDeclaration candidate
+where duplicatedAccess5(same1, same2, different1, different2, candidate)
+select same2,
+  "Duplicated variable '" + same1.getTarget().getName() +
+    "' may be an error, because '$@' might have been intended.", candidate, candidate.getName()
```
Comment (hvitved) [inline csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql]: I don't think we put `'` around `$@` placeholders.

File: csharp/ql/test/query-tests/Likely Bugs/PossiblyIncorrectVariable/PossiblyIncorrectVariable.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+namespace Test
+{
+    class Point1
+    {
+        public double x, y;
+
+        public Point1(double x, double y)
+        {
+            this.x = x;
+            this.x = y;  // this.x should be this.y.
```
Comment (hvitved) [inline csharp/ql/test/query-tests/Likely Bugs/PossiblyIncorrectVariable/PossiblyIncorrectVariable.cs]: Prefix all the bad cases with `BAD: `

File: csharp/ql/test/query-tests/Likely Bugs/PossiblyIncorrectVariable/PossiblyIncorrectVariable.cs [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        public Point1(double x, double y, double z)
+        {
+            this.x = x;
+            this.y = y;  // Good
```
Comment (hvitved) [inline csharp/ql/test/query-tests/Likely Bugs/PossiblyIncorrectVariable/PossiblyIncorrectVariable.cs]: Good -> GOOD (throughout)

**Pull request closed without merging**